scanpy/preprocessing/combat.py

```diff
@@ -0,0 +1,161 @@
+    n_array = float(sum(n_batches))
+
+    # drop intercept
+    drop_cols = [cname for cname, inter in  ((model == 1).all()).iteritems() if inter == True]
+    model = model[[c for c in model.columns if not c in drop_cols]]
+    design = design_mat(model, batch_levels)
+
+    # standardize across genes using a pooled variance estimator
+    sys.stderr.write("Standardizing Data across genes.\n")
+    B_hat = np.dot(np.dot(la.inv(np.dot(design.T, design)), design.T), data.T)
+    grand_mean = np.dot((n_batches / n_array).T, B_hat[:n_batch,:])
+    var_pooled = np.dot(((data - np.dot(design, B_hat).T)**2), np.ones((int(n_array), 1)) / int(n_array))
+    if np.sum(var_pooled == 0) > 0:
+        print('Found {} genes with zero variance.'.\
+          format(np.sum(var_pooled == 0)))
+    stand_mean = np.dot(grand_mean.T.reshape((len(grand_mean), 1)), np.ones((1, int(n_array))))
+    tmp = np.array(design.copy())
+    tmp[:,:n_batch] = 0
+    stand_mean  += np.dot(tmp, B_hat).T
+
+    # need to be a bit careful with the zero variance genes
+    s_data = np.where(var_pooled == 0, 0, \
+        ((data - stand_mean) / np.dot(np.sqrt(var_pooled), np.ones((1, int(n_array))))))
+    s_data = pd.DataFrame(s_data, index=data.index, columns=data.columns)
+
+    # fitting the parameters
+    sys.stderr.write("Fitting L/S model and finding priors\n")
+    batch_design = design[design.columns[:n_batch]]
+    gamma_hat = np.dot(np.dot(la.inv(np.dot(batch_design.T, batch_design)), batch_design.T), s_data.T)
+    delta_hat = []
+
+    for i, batch_idxs in enumerate(batch_info):
+        delta_hat.append(s_data[batch_idxs].var(axis=1))
+
+    gamma_bar = gamma_hat.mean(axis=1) 
+    t2 = gamma_hat.var(axis=1)
+    a_prior = list(map(aprior, delta_hat))
+    b_prior = list(map(bprior, delta_hat))
+
+    sys.stderr.write("Finding parametric adjustments\n")
+    gamma_star, delta_star = [], []
+    for i, batch_idxs in enumerate(batch_info):
+        temp = it_sol(s_data[batch_idxs], gamma_hat[i],
+                     delta_hat[i], gamma_bar[i], t2[i], a_prior[i], b_prior[i])
+
+        gamma_star.append(temp[0])
+        delta_star.append(temp[1])
+
+    sys.stdout.write("Adjusting data\n")
+    bayesdata = s_data
+    gamma_star = np.array(gamma_star)
+    delta_star = np.array(delta_star)
+
+
+    for j, batch_idxs in enumerate(batch_info):
+
+        dsq = np.sqrt(delta_star[j,:])
+        dsq = dsq.reshape((len(dsq), 1))
+        denom =  np.dot(dsq, np.ones((1, n_batches[j])))
+        numer = np.array(bayesdata[batch_idxs] - np.dot(batch_design.loc[batch_idxs], gamma_star).T)
+
+        bayesdata[batch_idxs] = numer / denom
+   
+    vpsq = np.sqrt(var_pooled).reshape((len(var_pooled), 1))
+    bayesdata = bayesdata * np.dot(vpsq, np.ones((1, int(n_array)))) + stand_mean
+    
+    # put back into the adata object
+    adata.X = bayesdata.values.transpose()
+ 
+
+def it_sol(sdat, g_hat, d_hat, g_bar, t2, a, b, conv=0.0001):
```

Comment:
Make this a private function. Also, at least some documentation would be good, otherwise no one will ever understand how this works.

Finally, this looks a lot like a function that can be dramatically be sped up using numba, eg as in `pp.calculate_qc_metrics`.

Reply:
Numba is a good option here but not entirely straightforward since data is stored in pd.DataFrame which numba doesn't know. Could add this in the future. I did however make it_sol a private function and I added some comments.

Reply:
you can access the underlying data using `some_series.values`: https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.values.html

That’s a numpy array that numba should understand.

Reply:
did that.